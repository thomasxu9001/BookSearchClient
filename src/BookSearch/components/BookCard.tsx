import React, {Component} from 'react';
import {RouteComponentProps, withRouter} from 'react-router';

import {Book} from '../Book';
import s from '../Book.less';


interface StateProps {
    item: Book
}

interface State {
}


export type Props = StateProps & RouteComponentProps;

export class BookCard extends Component<Props, State> {

    handleBookCardClick = () => {
        const {item} = this.props;
        this.props.history.push('/book/' + item.id);
    };

    render() {
        const {item} = this.props;
        return (
            <div className={s.bookCard} onClick={this.handleBookCardClick}>
                <div className={s.card__field}>
                    Title: {item.title}
                </div>
                <div className={s.card__field}>
                    Author: {item.author}
                </div>
            </div>
        );
    }
}

export default withRouter(BookCard);



