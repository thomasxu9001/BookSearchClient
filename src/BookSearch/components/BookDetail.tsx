import React, {Component} from 'react';

import {Book} from '../Book';
import s from "../Book.less";

interface StateProps {
    bookId: number
}

interface State {
    error: any
    isLoaded: boolean
    book: Book | any
}


export type Props = StateProps;

export class BookDetail extends Component<Props, State> {

    constructor(props: Props) {
        super(props);
        this.state = {
            error: null,
            isLoaded: false,
            book: []
        };
    }

    componentDidMount() {
        this.getBookDetail();
    };

    getBookDetail = () => {
        const {bookId} = this.props;
        const url = "http://localhost:8000/Rest/book/id/" + bookId;

        fetch(url, {
            method: 'GET'
        })
            .then(res => res.json())
            .then(
                (result) => {
                    this.setState({
                        isLoaded: true,
                        book: result
                    });
                },
                (error) => {
                    this.setState({
                        isLoaded: true,
                        error
                    });
                }
            );
    };

    render() {
        const {book} = this.state;
        return (
            <div>
                {book &&
                <div>
                    <div className={s.bookDetailTitle}>
                        {book.title}
                    </div>
                    <div className={s.bookDetailAuthor}>
                        Author: {book.author}
                    </div>
                    <div className={s.bookDetailDescription}>
                        {book.description}
                    </div>
                </div>
                }
            </div>
        );
    }
}

export default BookDetail;



