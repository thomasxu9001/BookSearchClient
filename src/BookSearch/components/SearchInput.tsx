import React, {Component} from 'react';
import {FontAwesomeIcon} from '@fortawesome/react-fontawesome';
import {debounce} from 'throttle-debounce';

import s from '../Book.less';
import {Book} from '../Book';

interface StateProps {
    searchBook: Function
}

interface State {
    searchHint: []
    searchValue: string
}


export type Props = StateProps;

export class SearchInput extends Component<Props, State> {
    private readonly debounceSearch: debounce<any>;

    constructor(props: Props) {
        super(props);
        this.debounceSearch = debounce(300, this.searchHint.bind(this));
        this.state = {
            searchHint: [],
            searchValue: ''
        };
    };

    onChangeHandler = (e: React.ChangeEvent<HTMLInputElement>) => {
        const {searchBook} = this.props;

        e.persist();
        // Update input value
        this.setState({searchValue: e.target.value});
        // Update search hint logic
        if (e.target.value && e.target.value.length > 0) {
           this.debounceSearch(e.target.value);
        } else {
            this.setState({searchHint: []}, () => searchBook());
        }

    };

    searchHint = (keyword?: string) => {
        // For better practise, we should have another api which just return matched title.
        const baseUrl = "http://localhost:8000/Rest/books";
        let url = keyword ? baseUrl + '?search=' + keyword : baseUrl;
        fetch(url, {
            method: 'GET'
        })
            .then(res => res.json())
            .then(
                (result) => {
                    this.setState({
                        searchHint: result['items']
                    });
                },
                (error) => {
                }
            );
    };

    searchClickHandler = () => {
        const {searchBook} = this.props;
        const {searchValue} = this.state;

        if (searchValue) {
            this.setState({searchHint: []}, () => searchBook(searchValue));
        }
    };

    dropdownClickHandler = (title: string) => {
        const {searchBook} = this.props;
        this.setState({searchHint: [], searchValue: title}, () => searchBook(title));
    };

    renderSearchHint = () => {
        const {searchHint} = this.state;

        let items = searchHint.map((item: Book) => {
                return <div className={s.hintItem} key={item.id}
                            onClick={() => this.dropdownClickHandler(item.title)}> {item.title}</div>
            }
        );

        return <div className={s.searchHintContainer}>
            {items}
        </div>
    };

    render() {
        const {searchHint, searchValue} = this.state;

        return (

            <div className={s.searchBoxContainer}>
                <div className={s.searchInputContainer}>
                    <input className={s.searchInput} onChange={this.onChangeHandler} value={searchValue}/>
                    <div className={s.searchIcon} onClick={this.searchClickHandler}>
                        <FontAwesomeIcon icon="search"/>
                    </div>

                </div>
                <div className={s.dropDownContainer}>
                    {searchHint && searchHint.length > 0 &&
                    this.renderSearchHint()
                    }
                </div>
            </div>

        );
    }
}

export default SearchInput;



