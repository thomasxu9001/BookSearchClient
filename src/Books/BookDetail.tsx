import React, {Component} from 'react';
import {Book} from "./Book";

interface StateProps {
    bookId: number
}

interface State {
    error: any
    isLoaded: boolean
    book: Book | any
}


export type Props = StateProps;

export class BookDetail extends Component<Props, State> {

    constructor(props: Props) {
        super(props);
        this.state = {
            error: null,
            isLoaded: false,
            book: []
        };
    }

    componentDidMount() {
        this.getBookDetail();
    };

    getBookDetail = () => {
        const {bookId} = this.props;
        const url = "http://localhost:8000/Rest/book/id/" + bookId;

        fetch(url, {
            method: 'GET'
        })
            .then(res => res.json())
            .then(
                (result) => {
                    this.setState({
                        isLoaded: true,
                        book: result
                    });
                },
                (error) => {
                    this.setState({
                        isLoaded: true,
                        error
                    });
                }
            );
    };

    render() {
        const {book} = this.state;
        return (
            <div>
                {book &&
                    <div>
                        <div className="bookDetailTitle">
                            {book.title}
                        </div>
                        <div className="bookDetailAuthor">
                            Author: {book.author}
                        </div>
                        <div className="bookDetailDescription">
                             {book.description}
                        </div>
                    </div>
                }
            </div>
        );
    }
}
export default BookDetail;



