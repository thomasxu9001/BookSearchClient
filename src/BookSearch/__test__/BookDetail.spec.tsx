import React from 'react';
import {mount, ReactWrapper} from 'enzyme';
import {BookDetail, Props} from '../components/BookDetail';


describe('BookDetail', () => {

    let
        subject: ReactWrapper<Props>,
        id: number
    ;

    beforeEach(() => {
        id = 1;
        subject = mount(
            <BookDetail bookId={id}/>
        );
    });

    it('looks good', async () => {
        expect(subject).toMatchSnapshot();
    });

});
