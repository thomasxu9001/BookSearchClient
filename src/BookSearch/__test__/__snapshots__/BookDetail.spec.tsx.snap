// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookDetail looks good 1`] = `ReactWrapper {}`;
